src/libexpr/eval.cc

```diff
@@ -1838,7 +1838,18 @@ string EvalState::copyPathToStore(PathSet & context, const Path & path)
     else {
         auto p = settings.readOnlyMode
             ? store->computeStorePathForPath(std::string(baseNameOf(path)), checkSourcePath(path)).first
-            : store->addToStore(std::string(baseNameOf(path)), checkSourcePath(path), FileIngestionMethod::Recursive, htSHA256, defaultPathFilter, repair);
+            : [&] {
+              // show paths we're copying, unless they're already in the store.
+              if (!store->isInStore(path)) {
+                auto path_to_add = checkSourcePath(path);
+                Activity act(*logger, lvlInfo, actCopyPath, fmt("copying path '%s'", path_to_add));
+                PushActivity pact(act.id);
+                return store->addToStore(std::string(baseNameOf(path)), path_to_add, FileIngestionMethod::Recursive, htSHA256, defaultPathFilter, repair);
+              }
+              else
+                return store->addToStore(std::string(baseNameOf(path)), checkSourcePath(path), FileIngestionMethod::Recursive, htSHA256, defaultPathFilter, repair);
+            } ();
```

Comment:
why not avoid duplication with?
```suggestion
            : [&] {
              // show paths we're copying, unless they're already in the store.
              auto path_to_add = checkSourcePath(path);
              if (!store->isInStore(path)) {
                Activity act(*logger, lvlInfo, actCopyPath, fmt("copying path '%s'", path_to_add));
                PushActivity pact(act.id);
              }
              return store->addToStore(std::string(baseNameOf(path)), path_to_add, FileIngestionMethod::Recursive, htSHA256, defaultPathFilter, repair);
            } ();
```
Also, maybe we need to include `repair` in the condition?

Reply:
That would be good, except as I understand it the Activity is supposed to exist while the addToStore is happening.  With this variation, the PushActivity goes out of scope before the addToStore call, and when it does it terminates the activity in its destructor.  This means the 'Activity' is over before the addToStore starts.

Reply:
Another possibility would be with a unique_ptr, like this:

```
        auto p = settings.readOnlyMode
            ? store->computeStorePathForPath(std::string(baseNameOf(path)), checkSourcePath(path)).first
            : [&] {
              // show paths we're copying, unless they're already in the store.
              std::unique_ptr<PushActivity> pact;
              auto path_to_add = checkSourcePath(path);
              if (!store->isInStore(path)) {
                Activity act(*logger, lvlInfo, actCopyPath, fmt("copying path '%s'", path_to_add));
                pact = std::unique_ptr<PushActivity>(new PushActivity(act.id));
              }
              return store->addToStore(std::string(baseNameOf(path)), path_to_add, FileIngestionMethod::Recursive, htSHA256, defaultPathFilter, repair);
            } ();
```

re 'repair' that makes sense to me.  I'd want to see repair paths being copied.  At present I don't know how to tell the difference between these and the fusillade of copyPath messages I got in the above example.

Reply:
took the unique_ptr approach to reduce duplication; added the repair flag.